arangod/Cluster/ActionBase.cpp

```diff
@@ -0,0 +1,253 @@
+
+inline static std::chrono::system_clock::duration secs_since_epoch() {
+  return std::chrono::system_clock::now().time_since_epoch();
+}
+
+ActionBase::ActionBase(MaintenanceFeature& feature, ActionDescription const& desc)
+  : _feature(feature), _description(desc), _state(READY),
+    _progress(0) {
+
+  init();
+
+}
+
+ActionBase::ActionBase(MaintenanceFeature& feature,  ActionDescription&& desc)
+  : _feature(feature), _description(std::move(desc)), _state(READY),
+    _progress(0) {
+
+  init();
+}
+
+void ActionBase::init() {
+  _hash = _description.hash();
+  _clientId = std::to_string(_hash);
+  _id = _feature.nextActionId();
+
+  // initialization of duration struct is not guaranteed in atomic form
+  _actionCreated = secs_since_epoch();
+  _actionStarted = std::chrono::system_clock::duration::zero();
+  _actionLastStat = std::chrono::system_clock::duration::zero();
+  _actionDone = std::chrono::system_clock::duration::zero();
+}
+
+
+ActionBase::~ActionBase() {
+}
+
+
+void ActionBase::notify() {
+
+  LOG_TOPIC(DEBUG, Logger::MAINTENANCE)
+    << "Job " << _description << " calling syncDBServerStatusQuo";
+  auto cf = ApplicationServer::getFeature<ClusterFeature>("Cluster");	
+  if (cf != nullptr) {	
+    cf->syncDBServerStatusQuo();	
+  }
+}
+
+
+/// @brief execution finished successfully or failed ... and race timer expired
+bool ActionBase::done() const {
+
+  return (COMPLETE==_state || FAILED==_state) &&
+    _actionDone.load() + std::chrono::seconds(_feature.getSecondsActionsBlock()) <= secs_since_epoch();
+
+} // ActionBase::done
+
+ActionDescription const& ActionBase::describe() const {
+  return _description;
+}
+
+MaintenanceFeature& ActionBase::feature() const {
+  return _feature;
+}
+
+VPackSlice const ActionBase::properties() const {
+  return _description.properties()->slice();
+}
+
+
+/// @brief Initiate a new action that will start immediately, pausing this action
+void ActionBase::createPreAction(std::shared_ptr<ActionDescription> const & description) {
+
+  _preAction = description;
+  std::shared_ptr<Action> new_action = _feature.preAction(description);
+
+  // shift from EXECUTING to WAITINGPRE ... EXECUTING is set to block other
+  //  workers from picking it up
+  if (_preAction && new_action->ok()) {
+    setState(WAITINGPRE);
+  } else {
+    _result.reset(TRI_ERROR_BAD_PARAMETER, "preAction rejected parameters.");
+  } // else
+
+} // ActionBase::createPreAction
+
+
+/// @brief Retrieve pointer to action that should run before this one
+std::shared_ptr<Action> ActionBase::getPreAction() {
+  return (_preAction != nullptr) ? _feature.findAction(_preAction) : nullptr;
+}
+
+
+/// @brief Retrieve pointer to action that should run after this one
+std::shared_ptr<Action> ActionBase::getPostAction() {
+  return (_postAction != nullptr) ? _feature.findAction(_postAction) : nullptr;
+}
+
+
+/// @brief Create a new action that will start after this action successfully completes
```

Comment:
Code path could corrupt registry object because this does hold lock.

Reply:
Have added a FIXMEMAINTENANCE